infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/msgHandlers/ReplicationHandlerMethods.java

```diff
@@ -0,0 +1,192 @@
+        return handler;
+    }
+
+    private void registerMethod(@Nonnull final MethodHandles.Lookup caller,
+                                @Nonnull final LogReplicationAbstractServer server,
+                                @Nonnull final Method method) {
+        final LogReplicationMsgHandler annotation = method.getAnnotation(LogReplicationMsgHandler.class);
+
+        // check if the method is already registered
+        if((!annotation.requestType().equals(RequestPayloadMsg.PayloadCase.NONE) && requestHandlerMap.containsKey(annotation.requestType())) ||
+                (!annotation.responseType().equals(ResponsePayloadMsg.PayloadCase.NONE) &&
+                responseHandlerMap.containsKey(annotation.responseType())))  {
+            throw new IllegalStateException("HandlerMethod for " + annotation.requestType() + " already registered!");
+        }
+
+        try {
+            HandlerMethod h;
+            if (Modifier.isStatic(method.getModifiers())) {
+                MethodHandle mh = caller.unreflect(method);
+                h = (HandlerMethod) LambdaMetafactory.metafactory(caller,
+                        "handle", MethodType.methodType(HandlerMethod.class),
+                        mh.type(), mh, mh.type()).getTarget().invokeExact();
+            } else {
+                // Instance method, so we need to capture the type.
+                MethodType mt = MethodType.methodType(method.getReturnType(), method.getParameterTypes());
+                MethodHandle mh = caller.findVirtual(server.getClass(), method.getName(), mt);
+
+                // Note: mtt captures the return type as well as the argument types of the annotated handler
+                // method. These should be consistent with the handle method described by the HandlerMethod
+                // functional interface.
+                MethodType mtt = mh.type().dropParameterTypes(0, 1);
+                h = (HandlerMethod) LambdaMetafactory.metafactory(caller, "handle",
+                        MethodType.methodType(HandlerMethod.class, server.getClass()),
+                        mtt, mh, mtt).getTarget().bindTo(server).invoke();
+            }
+
+            String payloadSting = annotation.requestType().equals(RequestPayloadMsg.PayloadCase.NONE) ?
+                    annotation.responseType().toString() : annotation.requestType().toString();
+
+            // Install pre-conditions on handler
+            final HandlerMethod handler = generateConditionalHandler(payloadSting, h);
+            // Install the handler in the map
+            if (!annotation.requestType().equals(RequestPayloadMsg.PayloadCase.NONE)) {
+                requestHandlerMap.put(annotation.requestType(), handler);
+            } else {
+                responseHandlerMap.put(annotation.responseType(), handler);
+            }
+        } catch (Throwable e) {
+            throw new UnrecoverableCorfuError("Exception during request handler registration", e);
+        }
```

Comment:
"request handler" -> "replication handler"? as it's for both request and response?